### PR TITLE
Catlearn relation rework

### PR DESCRIPTION
**Pull request by Previsou**
Took way more time than expected. I kept label_universe as a mapping instead of direct tensor because torch tensors are not hashable, and it was breaking many comparison, set retrieval, etc.

### CODE REVIEW

File: catlearn/categorical_model.py [hunk 1]
```diff
@@ -95,7 +92,8 @@ def __init__(
         """
         assert epsilon > 0., "epsilon should be strictly positive"
 
-        self._relation_models = relation_models
+        self._relation_model = relation_model
+        self._label_universe = label_universe
```
Comment (Kuraao) [inline catlearn/categorical_model.py:96]: add a property to access a MappingProxyType to label_universe (see relations property in previous version)

Comment (Previsou) [reply catlearn/categorical_model.py:96]: Done

**Comment by Previsou** [comment]
@Kuraao I also added the dual operator to DirectedGraph on this branch

File: catlearn/graph_utils.py [hunk 2]
```diff
@@ -69,6 +69,27 @@ def op(self) -> DirectedGraph[NodeType]:
         """
         return self.reverse(copy=False)
 
+    @property
+    def dual(self) -> DirectedGraph[NodeType]:
```
Comment (Kuraao) [inline catlearn/graph_utils.py:73]: I don't like the name, we should find an other one (it's clearly not a dual of the graph; although relations are dualized)

Maybe don't make it a property and rename it dualize_relations ?

File: catlearn/relation_cache.py [hunk 3]
```diff
@@ -154,6 +155,13 @@ def data(
         # if arrow has length 0, return data corresponding to its node
         return self._datas[arrow[0]]  # type: ignore
 
+    @property
+    def label_universe(self) -> Mapping[ArrowType, Tsor]:
```
Comment (Kuraao) [inline catlearn/relation_cache.py:159]: Add the same to DecisionCatModel (as it is there is no way to access it from the outside if I'm not mistaken).

**Comment by Previsou** [comment]
Added your suggestions

**Pull request merged**